### PR TITLE
Fix the race condition and make the alternate urls automatically roun…

### PR DESCRIPTION
**Pull request by schmidtw**
…d-robin along the way.

### CODE REVIEW

File: src/caduceus/outboundSender.go [hunk 1]
```diff
@@ -442,7 +471,7 @@ func (obs *CaduceusOutboundSender) dispatcher() {
 
 // worker is the routine that actually takes the queued messages and delivers
 // them to the listeners outside webpa
-func (obs *CaduceusOutboundSender) send(secret, acceptType string, msg *wrp.Message) {
+func (obs *CaduceusOutboundSender) send(urls *ring.Ring, secret, acceptType string, msg *wrp.Message) {
```
Comment (kcajmagic) [inline src/caduceus/outboundSender.go]: `urls` is part of the `CaduceusOutboundSender`. We don't need to change the method signature.

File: src/caduceus/outboundSender.go [hunk 2]
```diff
@@ -465,12 +494,12 @@ func (obs *CaduceusOutboundSender) send(secret, acceptType string, msg *wrp.Mess
 	}
 	payloadReader = bytes.NewReader(body)
 
-	req, err := http.NewRequest("POST", obs.id, payloadReader)
+	req, err := http.NewRequest("POST", urls.Value.(string), payloadReader)
```
Comment (kcajmagic) [inline src/caduceus/outboundSender.go]: For the first request, we should use the primary url (aka `id`). If that fails then we can leverage the alternative urls.
```suggestion
	req, err := http.NewRequest("POST", obs.id, payloadReader)
```

File: src/caduceus/outboundSender.go [hunk 3]
```diff
@@ -412,6 +437,10 @@ func (obs *CaduceusOutboundSender) dispatcher() {
 		obs.queueDepthGauge.Add(-1.0)
 
 		obs.mutex.RLock()
+		urls := obs.urls
+		// Move to the next URL to try 1st the next time.
+		obs.urls = obs.urls.Next()
```
Comment (kcajmagic) [inline src/caduceus/outboundSender.go]: I don't think this is needed.

File: src/caduceus/outboundSender.go (hunk 1 above)
Comment (schmidtw) [reply src/caduceus/outboundSender.go]: By sending the urls, we avoid the race condition where the `CaduceusOutboundSender` version could change, or needing to lock.  We don't want to lock if at all possible.

File: src/caduceus/outboundSender.go (hunk 3 above)
Comment (schmidtw) [reply src/caduceus/outboundSender.go]: This allows us to move on to the next url & load balance across all alternatives.

**Comment by schmidtw** [comment]
The nice thing about leveraging the alternatives for mainstream delivery, we provide our customers a mechanism for very uniform load balancing if they so desire.  I am thinking we just use the base url or the alternatives to make it simpler for callers of the api.  We'd still use the base url for the id in both cases.

**Pull request merged**